### PR TITLE
Case Review Tab Initial Draft; Minor Iterations

### PR DESCRIPTION
**Pull request by alexkb0009**
Should be ready for review.

- Allows selection of notes in "Case Review" tab and then saving them to Project for users with edit permission for all selected notes (but not edit permission for Variant or Gene).
  - Updates each Note's status to "current" and adding `interpretation`, `variant_notes`, and `discovery_interpretation` to Variant Items and `gene_note` and `discovery_interpretations` to Gene Item.
  - Uses new `/variant-samples/<variant-sample-uuid>/@@process-notes/` endpoint to accept PATCH `{ save_to_project_notes: { interpretation: uuid, discovery_intepretation: uuid, variant_notes: uuid, gene_notes: uuid } }` to then validate and make edits.
- Introduces a few new non-functional buttons like "Send to Report" which will be implemented in near future.
- Moved some stuff out of VariantSampleSelection into own file.
- Adopted more usage of `/embed` endpoint for performance improvements (i.e. loading VariantSampleList).
- Adjusted schema field names and added upgraders+test for them.
- Minor unrelated changes (annotated in comments below).
- _Backend:_ Moving definition of `request.user_info` (which just requests+caches `/session-properties`) to NamespacedAuthenticationPolicy from Auth0AuthenticationPolicy so that it is available within unit tests as well (and consistent with production/development Auth0 auth policy).
- _Backend:_ (Possibly needs exploration/review) changed `_auth0_await_user` to check for `@id` instead of `uuid`, because collection response started not returning UUIDs in test.

@sbreiff @drio18 @willronchetti @netsettler Please checkout/glance at the backend edits especially if have a moment please.
Will try to get this deployed to cgaptest soon and then post on Slack.

### CODE REVIEW

File: src/encoded/types/note.py [hunk 1]
```diff
@@ -42,7 +42,8 @@ def note_edit(context, request, render=None):
     next_status = request.validated.get("status")
     if next_status != previous_status and next_status == "current":
         request.validated["approved_date"] = datetime.datetime.utcnow().isoformat() + "+00:00"
-        request.validated["approved_by"] = request.user_info["details"]["uuid"]
+        auth_source, authid = request.authenticated_userid.split(".", 1)
```
Comment (alexkb0009) [inline src/encoded/types/note.py:45]: `request.user_info` was inaccessible within test context, as that's not based on Auth0. Considering whether makes sense to include user_info regardless of Auth0, perhaps elsewhere in authentication.py.

File: src/encoded/tests/test_auth0.py [hunk 2]
```diff
@@ -77,17 +77,17 @@ def auth0_4dn_user_profile():
 
 
 @Retry.retry_allowed(retries_allowed=20, wait_seconds=0.5)
-def _auth0_await_user(testapp, user_uuid):
+def _auth0_await_user(testapp, user_at_id):
```
Comment (alexkb0009) [inline src/encoded/tests/test_auth0.py:80]: [Error](https://gyazo.com/c69152d81eaa0129b695298ac57c61c5) starting occurring (where no "uuid" returned from search response of Users) after adding unit test for the @@process-notes endpoint. Not sure what cause might be, could use a double-check look into.

File: src/encoded/authentication.py [hunk 3]
```diff
@@ -137,13 +137,39 @@ def __init__(self, namespace, base, *args, **kw):
...
+        """
+        namespaced_userid = super(NamespacedAuthenticationPolicy, self) \
+            .authenticated_userid(request)
+
+        if namespaced_userid is not None:
+            # userid, if present, may be in form of UUID (if remoteuser) or an email (if Auth0).
+            namespace, userid = namespaced_userid.split(".", 1)
+
+            # Allow access basic user credentials from request obj after authenticating & saving request
+            def get_user_info(request):
```
Comment (alexkb0009) [inline src/encoded/authentication.py:159]: Moved this from `Auth0AuthenticationPolicy` class (defined lower in this file) to this `NamespacedAuthenticationPolicy` class. As I understand, NamespacedAuthenticationPolicy is used for all auth types and then 'delegates' out to Auth0 or other policy.

Reason for this is to be able to access `request.user_info` for any auth'd request (such as note_edit in note.py), whether it is as a remote user such as unit test or Auth0-authenticated user, for _consistency between production and testing_.

File: src/encoded/types/note.py (hunk 1 above)
Comment (alexkb0009) [reply src/encoded/types/note.py]: Reverted back to using `request.user_info` after making it accessible within unit tests by moving where is defined. This is because at this point (`note_edit` func), we expect it to be in form of UUID, and with `authenticated_userid` from Auth0AuthenticationPolicy, we get back email instead of UUID. `request.user_info` reliably supplies the user uuid.

But it's still a subrequest so maybe there's some faster of way of getting user UUID --or-- instead of overriding `item_edit` with `note_edit`, could use an 'upgrader' or something if that would work .. would need to find precedent example of an upgrader.

File: .gitignore [hunk 4]
```diff
@@ -37,6 +37,7 @@ beanstalk.cfg
 /src/encoded/static/scss/_variables.original.scss
 
 /src/encoded/tests/data/temp-local-inserts/*
+/src/encoded/tests/data/temp-local-inserts*
```
Comment (alexkb0009) [inline .gitignore:40]: I have a number of directories on my local environment starting with temp-local-inserts that are not in use at moment, e.g. `temp-local-inserts-2021-06-01`, `temp-local-inserts-2021-06-01-testing-something`, etc. so added this to handle such case. Can revert and look for other approaches to ignore those though, if desired.

File: src/encoded/custom_embed.py [hunk 5]
```diff
@@ -40,10 +40,11 @@ class CustomEmbed:
...
     def __init__(self, request, item, embed_props):
         self.request = request
-        self.ignored_embeds = embed_props["ignored_embeds"]
-        self.desired_embeds = embed_props["desired_embeds"]
-        self.embed_depth = embed_props["embed_depth"]
-        self.requested_fields = embed_props["requested_fields"]
+        self.ignored_embeds = embed_props.get("ignored_embeds", [])
+        self.desired_embeds = embed_props.get("desired_embeds", [])
+        self.embed_depth = embed_props.get("embed_depth", 4)
+        self.requested_fields = embed_props.get("requested_fields", [])
```
Comment (alexkb0009) [inline src/encoded/custom_embed.py:46]: @drio18 This change seemed safe since those defaults (like depth: 4) are passed in (but maybe not used if e.g. "fields" present) anyway down in the endpoint/view-config code, but could definitely use double-check.

File: src/encoded/static/components/index.js [hunk 6]
```diff
@@ -54,6 +52,7 @@ content_views.register(FileProcessedView,       'FileProcessed');
 content_views.register(SampleView,              'Sample');
 content_views.register(HealthView,              'Health');
 content_views.register(DocumentView,            'Document');
+content_views.register(DocumentView,            'Image');
```
Comment (alexkb0009) [inline src/encoded/static/components/index.js:55]: Unrelated edit made for an earlier demo -- reuses DocumentView for Image type Items.

File: src/encoded/static/components/item-pages/CaseView/CaseViewEmbeddedVariantSampleSearchTable.js [hunk 7]
```diff
@@ -17,10 +17,11 @@ export function CaseViewEmbeddedVariantSampleSearchTable(props){
     const {
         // Get/reuse default colExtMap from EmbeddedItemSearchTable
         columnExtensionMap: originalColExtMap = EmbeddedItemSearchTable.defaultProps.columnExtensionMap,
-        selectedItems,
-        onSelectItem,
+        selectedVariantSamples,
+        onSelectVariantSample,
```
Comment (alexkb0009) [inline src/encoded/static/components/item-pages/CaseView/CaseViewEmbeddedVariantSampleSearchTable.js:21]: Renamed some existing variables for explicitness and ease of searching (selectedItems is used relatively commonly, including in shared-portal-components).

File: src/encoded/renderers.py [hunk 8]
```diff
@@ -198,8 +198,6 @@ def security_tween(request):
                             # renderer.js. Is used to get access to User Info on initial web page render.
                             response.headers['X-Request-JWT'] = request.cookies.get('jwtToken', '')
                             user_info = request.user_info.copy()  # Re-ified property set in authentication.py
-                            # Redundant - don't need this in SSR nor browser as get from X-Request-JWT.
-                            del user_info["id_token"]
```
Comment (alexkb0009) [inline src/encoded/renderers.py:202]: We moved `request.user_info` and removed the token from being present in it in the first place (since it might occur for other authentication policies which do not have token).

As shown here, is straightforward to get the token from cookie (secure+httpOnly).

File: src/encoded/schemas/note.json [hunk 9]
```diff
@@ -34,7 +34,11 @@
             "title": "Approved By",
             "type": "string",
             "description": "The user who approved this note for a case",
-            "linkTo": "User"
+            "linkTo": "User",
+            "exclude_from": [
+                "FFedit-create"
+            ],
+            "permission": "restricted_fields"
```
Comment (alexkb0009) [inline src/encoded/schemas/note.json:41]: This field is set/updated by _backend only_ for security.

File: src/encoded/schemas/note.json [hunk 10]
```diff
@@ -47,7 +51,25 @@
...
+            "default": "in review",
+            "enum": [
+                "shared",
+                "obsolete",
+                "current",
+                "inactive",
+                "in review",
+                "deleted"
+            ],
+            "notes": "This is copy of mixins->status, but without permissions:restricted_fields to allow users to control."
```
Comment (alexkb0009) [inline src/encoded/schemas/note.json:72]: We can consider changing this to still be 'restricted_fields', and then having status change occur in our new `/@@process-notes/` endpoint (along with the approved_by and approved_date fields).

File: src/encoded/static/components/app.js [hunk 11]
```diff
@@ -964,28 +964,30 @@ export default class App extends React.PureComponent {
                     fallbackCallback(err);
                 }
 
-                console.error('Error in App.navigate():', err);
-
-                if (err.status === 500){
+                if (err.code === 500){
```
Comment (alexkb0009) [inline src/encoded/static/components/app.js:967]: Unrelated edits -- I think at some point Pyramid started returning 'code' (int) instead of 'status' (int).

File: src/encoded/schemas/note.json (hunk 10 above)
Comment (sbreiff) [reply src/encoded/schemas/note.json:72]: probably not a bad idea ^^

File: src/encoded/schemas/variant_sample_list.json [hunk 12]
```diff
@@ -64,12 +64,13 @@
                         "type": "string",
                         "description": "Stringified JSON representing the compound filter block request made when this Variant Sample was selected"
                     },
-                    "userid": {
-                        "title": "User ID",
+                    "selected_by": {
```
Comment (sbreiff) [inline src/encoded/schemas/variant_sample_list.json:67]: just be aware that changing the name of this field you will need to either A) write a schema upgrader or B) delete this field from all items in DB before it gets deployed.

File: src/encoded/types/base.py [hunk 13]
```diff
@@ -413,6 +413,16 @@ def create(context, request):
             'href': '{item_uri}?currentAction=create'.format(item_uri=request.resource_path(context)),
         }
 
+@snovault.calculated_property(context=Item, category='action')
+def create(context, request):
```
Comment (sbreiff) [inline src/encoded/types/base.py:417]: this looks like a duplicate of the above function - i think you forgot to change the name in the defline to `clone` perhaps?

Comment (alexkb0009) [reply src/encoded/types/base.py:417]: Thanks!

File: src/encoded/types/variant.py [hunk 14]
```diff
@@ -644,3 +624,263 @@ def download(context, request):
...
+        # 'variant_samples.variant_sample_item.discovery_interpretation.gene_candidacy',
+        # 'variant_samples.variant_sample_item.discovery_interpretation.variant_candidacy',
+        # 'variant_samples.variant_sample_item.associated_genotype_labels.proband_genotype_label',
+        # 'variant_samples.variant_sample_item.associated_genotype_labels.mother_genotype_label',
+        # 'variant_samples.variant_sample_item.associated_genotype_labels.father_genotype_label',
+        # 'structural_variant_samples.structural_variant_sample_item.structural_variant.display_title',
+        # 'structural_variant_samples.structural_variant_sample_item.interpretation.classification',
+        # 'structural_variant_samples.structural_variant_sample_item.discovery_interpretation.gene_candidacy',
+        # 'structural_variant_samples.structural_variant_sample_item.discovery_interpretation.variant_candidacy',
+    ]
```
Comment (sbreiff) [inline src/encoded/types/variant.py:867]: are these commented out because you're using the custom embed class instead?

Comment (alexkb0009) [reply src/encoded/types/variant.py:901]: Yep; I merged in master and it had a couple new fields here (`structural_variant_samples.` ...) so didn't want to lose that change completely, hence just commenting out. I'm not sure/up-to-date-on if VariantSampleList will have a `structural_variant_samples` though (?).

Comment (sbreiff) [reply src/encoded/types/variant.py:901]: it will have them, its a separate property than the SNVs.

File: src/encoded/types/note.py [hunk 15]
```diff
@@ -24,6 +26,10 @@ class Note(Item):
     embedded_list = []
 
 
+
```
Comment (drio18) [inline src/encoded/types/note.py:29]: As commented below for the VSL, consider removing some of this extra added white space.

File: src/encoded/types/variant.py [hunk 16]
```diff
@@ -644,3 +624,308 @@ def download(context, request):
...
+
+            ln[note_type_name] = loaded_note
+
+    for note_type_name in ["interpretation", "discovery_interpretation", "variant_notes", "gene_notes"]:
+        validate_and_load_note(note_type_name)
+
+    if len(ln) == 0:
+        raise HTTPBadRequest("No Note UUIDs supplied.")
+
+
```
Comment (drio18) [inline src/encoded/types/variant.py:676]: Throughout this view_config, consider deleting the large sections of white space between lines.

File: src/encoded/upgrade/variant.py [hunk 17]
```diff
@@ -18,3 +18,19 @@ def variant_1_2(value, system):
...
+    """
+    Update `variant_samples.userid` to `variant_samples.selected_by`.
+    We now can embed the user as linkto so simply calling it userid makes less sense.
+    We by default don't have `selected_by` in elasticsearch embedded_list, but can get
+    it when needed through the `/embed` API.
+    """
+
+    for vs_object in value.get("variant_samples", []):
+        if "userid" in vs_object:
+            vs_object["selected_by"] = vs_object["userid"]
```
Comment (drio18) [inline src/encoded/upgrade/variant.py:35]: I could be wrong here, but the value of `vs_object["userid"]` was a uuid for previous notes, and now they will be at_ids. Consider upgrading the values to be at_ids as well for consistency; if so, update the upgrade test as well to expect an at_id.

**Review by drio18 — COMMENTED** [review]
Looking forward to testing this out on cgaptest. I focused on the data model changes and the new endpoint, most of which looked good. No substantial/blocking comments, but I think more should be added to the endpoint test and the upgrader may need a small change.

File: src/encoded/static/components/item-pages/VariantSampleView/InterpretationSpaceController.js [hunk 18]
```diff
@@ -918,13 +918,17 @@ function GenericInterpretationSubmitButton(props) {
...
 
-    function discardAndNavigate() {
+    const onHide = useCallback(function(){
+        setIsSubmitting(false, null, false);
+    });
+
+    const discardAndNavigate = useCallback(function(){
         setIsSubmitting(false, () => navigate(href, { 'skipRequest' : false, 'replace': false }), false);
-    }
+    });
```
Comment (Bianca-Morris) [inline src/encoded/static/components/item-pages/VariantSampleView/InterpretationSpaceController.js:927]: Just curious for future reference what the benefit is in using useCallback in these sort of instances?

**Review by Bianca-Morris — COMMENTED** [review]
Code looks good; I'll take a second look/do a second round of testing when you fix the `user_info` bug we discussed. I wrote more detailed steps-to-reproduce in one of the comments here, so if you're having trouble reproducing you'll know exactly which case and vsl I pulled to encounter it.

File: src/encoded/authentication.py (hunk 3 above)
Comment (alexkb0009) [reply src/encoded/authentication.py:159]: Resolved (was an earlier version of branch)

File: src/encoded/schemas/note.json (hunk 10 above)
Comment (alexkb0009) [reply src/encoded/schemas/note.json]: (All set)

File: src/encoded/types/variant.py (hunk 16 above)
Comment (alexkb0009) [reply src/encoded/types/variant.py:679]: The whitespace is to help organize the code better (sometimes / when it makes sense) and break apart the large procedural func into separate comprehensible parts.

File: src/encoded/upgrade/variant.py (hunk 17 above)
Comment (alexkb0009) [reply src/encoded/upgrade/variant.py:35]: I have no idea; for some reason it was giving me back at_id instead of uuid in an unrelated test as well as here so I used at_id (and switched the unrelated test to using at_id on other 1 as a quickfix as couldn't figure out why it changed). Pls feel free to adjust any of the unit tests re: using proper identifiers.

Comment (alexkb0009) [reply src/encoded/upgrade/variant.py:35]: This field was changed to be a literal linkTo so I'm not sure it matters for cases of PATCHing/POSTing and upgraders, i.e. if submitted as UUID or AtID before saving. In either case it becomes a linkTo during indexing. An will have an AtID if call `@@object` frame

File: src/encoded/types/variant.py (hunk 16 above)
Comment (alexkb0009) [reply src/encoded/types/variant.py:679]: But yeah I'll try to reduce it down a bit.. previously was a bunch more logic until added re-usable funcs and had the "approved_by" and "date_approved" be generated in this func so is less sense now to have 4-line whitespace delineation.

File: src/encoded/tests/test_types_variant_sample.py [hunk 19]
```diff
@@ -91,6 +110,59 @@ def test_variant_sample_list_post(bgm_user_testapp, variant_sample_list1):
     bgm_user_testapp.post_json('/variant_sample_list', variant_sample_list1, status=201)
 
 
+def test_variant_sample_patch_notes_process_success(bgm_user, testapp, bgm_user_testapp, bgm_test_variant_sample, bgm_test_variant_sample_for_note_patch_process, bgm_test_variant_sample_for_note_patch_process2):
```
[the remark this answers is not in the document]
Comment (willronchetti) [reply src/encoded/tests/test_types_variant_sample.py:113]: Agreeing with this here, we should cover at least these cases. Happy to help out if needed.

File: src/encoded/types/variant.py [hunk 20]
```diff
@@ -647,3 +627,308 @@ def download(context, request):
...
+    """
+    Accepts (may be extended in future):
+        {
+            "save_to_project_notes" : {
+                "variant_notes": UUID,
+                "gene_notes": UUID,
+                "interpretation": UUID,
+                "discovery_interpretation": UUID,
+            }
+        }
```
Comment (willronchetti) [inline src/encoded/types/variant.py:648]: This is good to know, but more info on the route itself should be in the docstring as well.

File: src/encoded/types/variant.py (hunk 16 above)
Comment (willronchetti) [reply src/encoded/types/variant.py:679]: If you're going to do this, 1 blank line is fine followed by a comment line ie:

```
if len(ln) == 0:
        raise HTTPBadRequest("No Note UUIDs supplied.")

# Patch Paylods
variant_patch_payload = {} # Can be converted to dict of variants if need to PATCH multiple in future
genes_patch_payloads = {} # Keyed by @id, along with `note_patch_payloads`
note_patch_payloads = {}
```

In addition, if you're doing this a lot, odds are it means you want multiple functions (or even a class).

File: src/encoded/types/variant.py [hunk 21]
```diff
@@ -647,3 +627,308 @@ def download(context, request):
...
+    note_patch_payloads = {}
+
+
+
+
+    need_variant_patch = "interpretation" in stpn \
+        or "discovery_interpretation" in stpn \
+        or "variant_notes" in stpn
+    need_gene_patch = "discovery_interpretation" in stpn \
+        or "gene_notes" in stpn
```
Comment (willronchetti) [inline src/encoded/types/variant.py:693]: Strongly prefer parens over `\` line extensions ie:

```
need_variant_patch = ("interpretation" in stpn 
                                       or "discovery_interpretation" in stpn 
                                       or "variant_notes" in stpn)
```

File: src/encoded/upgrade/note.py [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+def note_discovery_1_2(value, system):
+    common_note_1_2_upgrade(value, system)
+
+@upgrade_step('note_interpretation', '1', '2')
+def note_interpretation_1_2(value, system):
+    common_note_1_2_upgrade(value, system)
+
+@upgrade_step('note_standard', '1', '2')
+def note_standard_1_2(value, system):
+    common_note_1_2_upgrade(value, system)
```
Comment (willronchetti) [inline src/encoded/upgrade/note.py:29]: FYI, upgraders are broken and do not function (or at least they didn't when I tried to upgrade `variant`). I haven't had time to look into why, but it must be an issue in snovault I assume.

File: src/encoded/upgrade/variant.py (hunk 17 above)
Comment (willronchetti) [reply src/encoded/upgrade/variant.py:35]: Same issue with upgraders. Did you find that these ran on cgaptest?

**Review by willronchetti — APPROVED** [review]
No major blockers but a few comments worth looking at that others have noted as well.

File: src/encoded/upgrade/note.py (hunk 22 above)
Comment (alexkb0009) [reply src/encoded/upgrade/note.py:29]: Thanks! Yeah I didn't really run locally (aside from tests) and was trying to use cgaptest as testing ground for it.. will dbl check some more.

Comment (alexkb0009) [reply src/encoded/upgrade/note.py:29]: Right now am getting errors for loading note, but it seems variant_sample_list upgrader may be working? Or at least not failing? Or maybe not being ran at all.. will give it a quick check.

Comment (alexkb0009) [reply src/encoded/upgrade/note.py:29]: Ah I forgot to add institution/project to the insert. It seems that my temp-local-inserts (which have set to be schema_version: "1") to have went thru upgrader, will see if got upgraded correctly.

File: src/encoded/upgrade/variant.py (hunk 17 above)
Comment (alexkb0009) [reply src/encoded/upgrade/variant.py:35]: It actually [seems it worked](https://gyazo.com/0cfe0cd65307e4392d761db1453115e0) (VSL from 7/13 has been upgraded from `userid` to `selected_by`), but am having trouble with processing inserts on local using it ... but think is OK? (i.e. we just manually upgrade our inserts if need be..)

File: src/encoded/authentication.py [hunk 23]
```diff
@@ -137,13 +137,39 @@ def __init__(self, namespace, base, *args, **kw):
...
             userid = self._namespace_prefix + userid
         return userid
 
+    def authenticated_userid(self, request):
+        """
+        Adds `request.user_info` for all authentication types.
+        Fetches and returns some user details if called.
+        """
+        namespaced_userid = super(NamespacedAuthenticationPolicy, self) \
+            .authenticated_userid(request)
```
[the remark this answers is not in the document]
Comment (netsettler) [reply src/encoded/authentication.py:152]: If I had to break it over lines, I would write it as
```
namespaced_userid = (super(NamespacedAuthenticationPolicy, self)
                     .authenticated_userid(request))
```
but in Python 3, you can just leave out the args to `super` in the common case of calling the parent class. This avoids problems if you rename the class, and it's also shorter. So please be sure to test this, but:
```
namespaced_userid = super().authenticated_userid(request)
```

File: src/encoded/upgrade/note.py (hunk 22 above)
Comment (alexkb0009) [reply src/encoded/upgrade/note.py:29]: Ultimately didn't get local inserts to run through upgrader but the fields on cgaptest seemed to have successfully upgraded when it was deployed, so considering as resolved (in context of this branch).
